### PR TITLE
add quandl as a source for historical financial data

### PR DESCRIPTION
**Pull request by briancappello**
Replaces https://github.com/pydata/pandas/pull/9935

| Q |	A |
------------ | -----
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Tests pass? | yes
| Fixed tickets | -
| License | BSD
| Doc PR | -

### CODE REVIEW

File: pandas_datareader/tests/test_data.py [hunk 1]
```diff
@@ -135,6 +135,95 @@ def test_unicode_date(self):
         self.assertEquals(data.index.name, 'Date')
 
 
+
+class TestQuandl(tm.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        print('hiiiiii')
```
Comment (dstephens99) [inline pandas_datareader/tests/test_data.py]: Should drop this line..

**Comment by dstephens99** [comment]
This looks great to me.  Can you add to the docs?

**Comment by davidastephens** [comment]
@briancappello Just following up on this pull-request.  Can you drop the print('hiiii') line in the tests and add this to the docs, then we can merge?

**Comment by femtotrader** [comment]
What about http://pythonhosted.org/Quandl 

moreover it seems that API v1 will be deprecated

https://www.quandl.com/help/api

(Note: The guide to v1 of the Quandl API is available here. We encourage users to adopt the current version, v1 is on the path to deprecation.)

**Comment by briancappello** [comment]
@davidastephens Apologies on dropping the ball here. I just pushed up some commits to address the outstanding issues raised; the debug print statement has been removed, I updated to using the quandl api v3, and added basic documentation.

**Comment by davidastephens** [comment]
Thanks @briancappello.  Can you rebase and squash?

@bashtage @femtotrader   Are you guys good with this?  Seems like a reasonable way to provide a consistent API between yahoo, google and quandl.

**Comment by femtotrader** [comment]
We shouldn't merge

see https://github.com/briancappello/pandas-datareader/tree/quandl_historical_data_source/pandas_datareader

all is in `data.py`

**Comment by femtotrader** [comment]
I also wonder differences with https://github.com/quandl/quandl-python/tree/feature-AP-243/v3-client

**Comment by briancappello** [comment]
@davidastephens @femtotrader I refactored this to match the new architecture. Tests are failing because quandl limits non-authenticated users to 50 API calls per day (there are less than 50 calls but I had pushed maybe an hour ago causing this run to fail). Does nose have something akin to "try to run this test but skip it if it fails" or maybe there's a better way to handle this class of error?

@femtotrader I haven't played with the official quandl package much; at first glance it seems to lack support for downloading more than one symbol at once, which we support here.

**Comment by bashtage** [comment]
You can tell nose that it is ok if a test fail.An example of a decorator that implements this is here: 

http://stackoverflow.com/questions/9613932/nose-plugin-for-expected-failures

50 API calls/day is a problem for testing IMO.  You should probably get an API key from quandl and follow 

http://docs.travis-ci.com/user/encryption-keys/

to insert an encrypted version into the main.  This will at least allow the test to always run on travis.

**Comment by bashtage** [comment]
As a comment I don't like the setting your API key as an environmental variable.  Why not let the api key be set at the module level and kept in a module level variable?  This would require a dedicated function like `set_quandl_api_key('xxxx')`

**Comment by briancappello** [comment]
@bashtage Thanks for the feedback! I added a set_api_key function to do (I think?) what you're talking about. As for the 50 calls/day being a problem, I agree with you there. Reading through the travis doc you linked, it mentions encrypted environment variables are not available for external pull requests (like, for example, this one or any other coming from a fork). I'm not very familiar with travis but it doesn't seem to support an approach other than environment variables for encrypted data?

**Comment by femtotrader** [comment]
@bashtage, @briancappello is right.

With Travis, we need to be able to set API key using an environnement variable.

But code should be able to support both
- environnement variable
- passing an API key parameter

We should probably have a parameter called `credentials`.
This parameter could be a dict and `api_key` could be one of its keys.

Quandl only need an API key but some other datareader could need `login` / `password`...

so this `credentials` parameter could also allow this:

    start = datetime(2010, 1, 1)
    end = datetime(2013, 1, 27)
    df = web.DataReader('F', 'quandl', start, end, credentials=dict(api_key='....'))

or if we want a more oriented object approach

    df = web.DataReader('F', 'quandl', start, end, credentials=Credential(api_key='....'))

but if `credendials` is `None` or if no `'api_key'` is found, environnement variable should be use.

If no environnement variable for Quandl is found, query will be performed without setting API key.

**Comment by femtotrader** [comment]
An other approach could be to have

    web.set_credentials(api_key='....')

so it could also allow (for other paid datareader)

    web.set_credendials(login='...', password='...')

but anyway I won't expose like you did a function like

    web.set_quandl_api_key('your_key_here')

because I think we should avoid to expose specific datareader function at this level.

**Comment by davidastephens** [comment]
Did we ever get anywhere on adding the API key? @femtotrader @briancappello @bashtage

**Comment by femtotrader** [comment]
I think we should close this PR. There is a Quandl Python library which is able to output Pandas DataFrame. see https://github.com/quandl/quandl-python and https://github.com/quandl/quandl-python/tree/feature-AP-243/v3-client2

**Pull request closed without merging**